torchrl/objectives/sac.py

```diff
@@ -1060,159 +1066,151 @@ def in_keys(self, values):
+            tensordict_reshape = tensordict.reshape(-1)
+        else:
+            tensordict_reshape = tensordict
 
-        tensordict_actor_grad = tensordict_select.select(
-            *obs_keys
-        )  # to avoid overwriting keys
-        next_td_actor = step_mdp(tensordict_select).select(
-            *self.actor_network.in_keys
-        )  # next_observation ->
-        tensordict_actor = torch.stack([tensordict_actor_grad, next_td_actor], 0)
-        tensordict_actor = tensordict_actor.contiguous()
+        loss_value, metadata_value = self._value_loss(tensordict_reshape)
+        loss_actor, _ = self._actor_loss(tensordict_reshape)
+        loss_alpha = self._alpha_loss(tensordict_reshape)
 
-        with set_exploration_type(ExplorationType.RANDOM):
-            # vmap doesn't support sampling, so we take it out from the vmap
-            td_params = self._vmap_getdist(
-                tensordict_actor,
-                actor_params,
+        tensordict_reshape.set(self.tensor_keys.priority, metadata_value["td_error"])
+        if loss_actor.shape != loss_value.shape:
+            raise RuntimeError(
+                f"Losses shape mismatch: {loss_actor.shape}, and {loss_value.shape}"
             )
-            if isinstance(self.actor_network, ProbabilisticActor):
-                tensordict_actor_dist = self.actor_network.build_dist_from_params(
-                    td_params
-                )
-            else:
-                tensordict_actor_dist = self.actor_network.build_dist_from_params(
-                    td_params
-                )
-            probs = tensordict_actor_dist.probs
-            z = (probs == 0.0).float() * 1e-8
-            logp_pi = torch.log(probs + z)
-            logp_pi_pol = torch.sum(probs * logp_pi, dim=-1, keepdim=True)
-
-        # repeat tensordict_actor to match the qvalue size
-        _actor_loss_td = (
-            tensordict_actor[0]
-            .select(*self.qvalue_network.in_keys)
-            .expand(self.num_qvalue_nets, *tensordict_actor[0].batch_size)
-        )  # for actor loss
-        _qval_td = tensordict_select.select(*self.qvalue_network.in_keys).expand(
-            self.num_qvalue_nets,
-            *tensordict_select.select(*self.qvalue_network.in_keys).batch_size,
-        )  # for qvalue loss
-        _next_val_td = (
-            tensordict_actor[1]
-            .select(*self.qvalue_network.in_keys)
-            .expand(self.num_qvalue_nets, *tensordict_actor[1].batch_size)
-        )  # for next value estimation
-        tensordict_qval = torch.cat(
-            [
-                _actor_loss_td,
-                _next_val_td,
-                _qval_td,
-            ],
-            0,
-        )
+        if shape:
+            tensordict.update(tensordict_reshape.view(shape))
+        out = {
+            "loss_actor": loss_actor.mean(),
+            "loss_qvalue": loss_value.mean(),
+            "loss_alpha": loss_alpha.mean(),
+            "alpha": self._alpha,
```

Comment:
shouldn't we detach?

Reply:
this is a property, if you see it should be already detached (i follow what you wrote in the other sac)
```

   @property
    def _alpha(self):
        if self.min_log_alpha is not None:
            self.log_alpha.data.clamp_(self.min_log_alpha, self.max_log_alpha)
        with torch.no_grad():
            alpha = self.log_alpha.exp()
        return alpha
```